### PR TITLE
MGMT-10411: Allow enabling virtual interfaces

### PR DESCRIPTION
**Pull request by CrystalChun**
[MGMT-10411](https://issues.redhat.com/browse/MGMT-10411)
As part of epic https://issues.redhat.com/browse/MGMT-10087
Exposes an environment variable in the service
that allows virtual interfaces to be
used.

## List all the issues related to this PR

- [x] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [x] assisted-test-infra environment: deployed and installed nodes after using this service image and adding `ENABLE_VIRTUAL_INTERFACES` to the configmap
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @carbonin 
/cc @filanov 
/cc @ori-amizur 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by carbonin** [comment]
Is this allowing the interfaces to be used or just indicating that the agent should collect and send them?

**Comment by CrystalChun** [comment]
> Is this allowing the interfaces to be used or just indicating that the agent should collect and send them?

It's intended for both, should they be separate?

File: cmd/main.go [hunk 1]
```diff
@@ -146,6 +146,7 @@ var Options struct {
 	HTTPListenPort                 string        `envconfig:"HTTP_LISTEN_PORT" default:""`
 	AllowConvergedFlow             bool          `envconfig:"ALLOW_COVERGED_FLOW" default:"true"`
 	IronicIgnitionBuilderConfig    ignition.IronicIgniotionBuilderConfig
+	EnableVirtualInterfaces        bool `envconfig:"ENABLE_VIRTUAL_INTERFACES" default:"false"`
```
[the remark this answers is not in the document]
Comment (CrystalChun) [reply cmd/main.go:149]: There was [discussion in the epic](https://issues.redhat.com/browse/MGMT-10087?focusedCommentId=20333260&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-20333260) that it should be globally enabled/disabled as opposed to at the cluster/infra-env level.

File: internal/host/hostcommands/inventory_cmd.go [hunk 2]
```diff
@@ -26,6 +29,9 @@ func (h *inventoryCmd) GetSteps(ctx context.Context, host *models.Host) ([]*mode
 			host.ID.String(),
 		},
 	}
+	if h.enableVirtualInterfaces {
```
Comment (filanov) [inline internal/host/hostcommands/inventory_cmd.go:32]: why do we need to send it to the agent? 
I was thinking that the agent will send everything and the service will decide if it's need to be used or not.

Comment (CrystalChun) [reply internal/host/hostcommands/inventory_cmd.go:32]: For backwards compatibility - if the agent is newer than the service and sends all the interfaces, the service will use the first interface (hard-coded) which is not always a useable one. 

Therefore we needed a way for the new agent to tell when the service is new enough to handle all the interfaces being sent.

Slack discussion for context: https://coreos.slack.com/archives/C014N2VLTQE/p1654481131218899

Comment (filanov) [reply internal/host/hostcommands/inventory_cmd.go:32]: i suspect that this is not backward compatible, we can have a case of new service old agent not the other way around

File: internal/host/host.go [hunk 3]
```diff
@@ -349,6 +350,25 @@ func (m *Manager) updateInventory(ctx context.Context, cluster *common.Cluster,
 		log.WithError(err).Errorf("not updating inventory - failed to check disks eligibility for host %s", h.ID)
 		return common.NewApiError(http.StatusInternalServerError, err)
 	}
+
+	// MGMT-10087: All interfaces will be sent to the service from the agent
```
[the remark this answers is not in the document]
Comment (CrystalChun) [reply internal/host/host.go]: Gotcha, removed it, thank you!

File: internal/common/test_configuration.go [hunk 4]
```diff
@@ -239,6 +239,41 @@ func GenerateTestDefaultInventory() string {
 	return string(b)
 }
 
+func GenerateTestInventoryWithVirtualInterface() string {
+	inventory := &models.Inventory{
+		Interfaces: []*models.Interface{
+			{
+				Name: "eth0",
+				IPV4Addresses: []string{
+					"1.2.3.4/24",
```
Comment (carbonin) [inline internal/common/test_configuration.go:248]: nit, but you could use addresses from the subnets specifically designated for use in tests and documentation here.

They are `192.0.2.0/24`, `198.51.100.0/24`, and `203.0.113.0/24`
https://en.wikipedia.org/wiki/Reserved_IP_addresses

File: internal/host/host_test.go [hunk 5]
```diff
@@ -1323,6 +1323,72 @@ var _ = Describe("UpdateInventory", func() {
...
+				nil, createValidatorCfg(), nil, defaultConfig, dummy, nil, nil)
+			Expect(enabled_hapi.UpdateInventory(ctx, &host, host.Inventory)).ToNot(HaveOccurred())
+
+			h := hostutil.GetHostFromDB(hostId, infraEnvId, db)
+			Expect(h.Inventory).ToNot(BeEmpty())
+			inventory, err := common.UnmarshalInventory(h.Inventory)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(inventory).To(Not(BeNil()))
+			Expect(inventory.Interfaces).Should(HaveLen(2))
+			Expect(inventory.Interfaces[0].Type).To(Equal(physicalInterface.Type))
```
Comment (carbonin) [inline internal/host/host_test.go:1388]: Can we rely on the ordering here? Might cause a flake if we can't. Maybe it's better to range through them and ensure you find the virtual interface?

File: internal/host/host_test.go [hunk 6]
```diff
@@ -1323,6 +1323,72 @@ var _ = Describe("UpdateInventory", func() {
...
+			Expect(err).ToNot(HaveOccurred())
+			Expect(inventory).To(Not(BeNil()))
+			Expect(inventory.Interfaces).Should(HaveLen(1))
+			Expect(inventory.Interfaces[0].Type).To(Equal(physicalInterface.Type))
+		})
+		It("Saves virtual interfaces when virtual interface flag is enabled", func() {
+			defaultConfig.EnableVirtualInterfaces = true
+			enabled_hapi := NewManager(common.GetTestLog(), db, mockEvents, mockValidator,
+				nil, createValidatorCfg(), nil, defaultConfig, dummy, nil, nil)
+			Expect(enabled_hapi.UpdateInventory(ctx, &host, host.Inventory)).ToNot(HaveOccurred())
```
Comment (carbonin) [inline internal/host/host_test.go:1380]: nit, but when I see `ToNot(HaveOccurred())` I always think we're expecting a method not to be called.

Personally I prefer `Expect(enabled_hapi.UpdateInventory(ctx, &host, host.Inventory)).To(Succeed())` in cases where we're asserting on the error response of a function that returns just an error value.

File: internal/host/host.go [hunk 7]
```diff
@@ -349,6 +350,22 @@ func (m *Manager) updateInventory(ctx context.Context, cluster *common.Cluster,
...
 		return common.NewApiError(http.StatusInternalServerError, err)
 	}
+
+	// Remove virtual interfaces if the feature is not enabled
+	if !m.Config.EnableVirtualInterfaces {
+		var physicalInterfaces []*models.Interface = make([]*models.Interface, 0)
+		for _, intf := range inventory.Interfaces {
+			if intf.Type == "" {
+				// Indicates an older agent, so all interfaces are non-virtual
+				break
```
Comment (carbonin) [inline internal/host/host.go:360]: If we break the loop here, when the (presumably) first interface doesn't have a type, won't we set `inventory.Interfaces` to the empty `physicalInterfaces` slice?

I think it would be better if you just treat the empty case the same as the other physical ones below and add it.

Also can you add a test case for an empty type?

File: internal/host/host_test.go [hunk 8]
```diff
@@ -1323,6 +1323,72 @@ var _ = Describe("UpdateInventory", func() {
 			Expect(h.InstallationDiskID).To(Equal(diskId))
 		})
 	})
+	Context("Interfaces", func() {
+		const (
+			diskName = "FirstDisk"
+			diskId   = "/dev/disk/by-id/FirstDisk"
+		)
+		var (
+			physicalInterface = models.Interface{
```
Comment (carbonin) [inline internal/host/host_test.go:1332]: This doesn't look like it is used anywhere other than where we assert on the type.
Couldn't we just hardcode "physical" in the test like we're doing everywhere else? Why define the whole structure?

File: internal/host/host_test.go (hunk 6 above)
Comment (CrystalChun) [reply internal/host/host_test.go]: Gotcha, modified it, thank you!

File: internal/host/host_test.go (hunk 5 above)
Comment (CrystalChun) [reply internal/host/host_test.go]: True, modified it to loop through the interfaces instead. Thank you!

File: internal/host/host_test.go (hunk 8 above)
Comment (CrystalChun) [reply internal/host/host_test.go]: Removed it, thank you!

File: internal/host/host.go (hunk 7 above)
Comment (CrystalChun) [reply internal/host/host.go]: That's right, removed this break and added tests for the empty type. Thank you!

File: internal/common/test_configuration.go (hunk 4 above)
Comment (CrystalChun) [reply internal/common/test_configuration.go]: Modified it! Thank you!

**Review by carbonin — APPROVED** [review]
